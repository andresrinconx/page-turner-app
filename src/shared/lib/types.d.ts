import { COLORS } from "./constants";
import type { ViewStyle, TextStyle } from "react-native";

declare global {
  type FlexProps = {
    flex?: number;
    flexDirection?: ViewStyle["flexDirection"];
    justifyContent?: ViewStyle["justifyContent"];
    alignItems?: ViewStyle["alignItems"];
    flexWrap?: ViewStyle["flexWrap"];
    gap?: number;
  };

  type SpacingProps = {
    p?: number;
    px?: number;
    py?: number;
    pt?: number;
    pb?: number;
    pl?: number;
    pr?: number;
    m?: number;
    mx?: number;
    my?: number;
    mt?: number;
    mb?: number;
    ml?: number;
    mr?: number;
  };

  type PositionProps = {
    position?: ViewStyle["position"];
    top?: number;
    bottom?: number;
    left?: number;
    right?: number;
  };

  type SizingProps = {
    w?: number | string;
    h?: number | string;
    minW?: number | string;
    minH?: number | string;
    maxW?: number | string;
    maxH?: number | string;
  };

  type ColorProps = {
    bg?: keyof typeof COLORS;
    color?: keyof typeof COLORS;
  };

  type TypographyProps = {
    fontSize?: number;
    lineHeight?: number;
    fontWeight?: TextStyle["fontWeight"];
    textAlign?: TextStyle["textAlign"];
    fontFamily?: string;
    letterSpacing?: number;
    textDecorationLine?: TextStyle["textDecorationLine"];
    textTransform?: TextStyle["textTransform"];
  };

  type BorderProps = {
    borderWidth?: number;
    borderTopWidth?: number;
    borderBottomWidth?: number;
    borderLeftWidth?: number;
    borderRightWidth?: number;
    borderRadius?: number;
    borderTopLeftRadius?: number;
    borderTopRightRadius?: number;
    borderBottomLeftRadius?: number;
    borderBottomRightRadius?: number;
    borderStyle?: ViewStyle["borderStyle"];
    borderColor?: keyof typeof COLORS;
  };

  type CommonStyleProps = FlexProps &
    SpacingProps &
    SizingProps &
    PositionProps &
    ColorProps &
    TypographyProps &
    BorderProps;
}

export {};
