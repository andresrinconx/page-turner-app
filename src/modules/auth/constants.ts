/**
 * constants.ts
 * Constants for the auth module
 *
 * Created by Andres Rincon on 21/4/25.
 */
