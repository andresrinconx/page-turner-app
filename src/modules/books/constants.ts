/**
 * constants.ts
 * Constants for the books module
 *
 * Created by Andres Rincon on 21/4/25.
 */
