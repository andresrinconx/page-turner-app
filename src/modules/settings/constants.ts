/**
 * constants.ts
 * Settings module constants
 *
 * Created by Andres Rincon on 28/4/25.
 */

import { WeekdayItem } from "./types";

export const DEFAULT_WEEKDAYS: WeekdayItem[] = [
  { id: 1, label: "M", selected: true },
  { id: 2, label: "T", selected: true },
  { id: 3, label: "W", selected: true },
  { id: 4, label: "T", selected: true },
  { id: 5, label: "F", selected: true },
  { id: 6, label: "S", selected: true },
  { id: 7, label: "S", selected: true },
];

export const STORAGE_KEYS = {
  WEEKDAYS: "notifications_weekdays",
  START_DATE: "notifications_start_date",
  REMINDER_TIME: "notifications_reminder_time",
};

export const NOTIFICATION_DEFAULTS = {
  TITLE: "Page Turner",
  BODY: "Time to read your book today!",
  HOUR: 8,
  MINUTE: 0,
};
