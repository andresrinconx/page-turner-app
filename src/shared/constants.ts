/**
 * constants.ts
 * Constants for the shared components
 *
 * Created by Andres Rincon on 21/4/25.
 */

export const COLORS = {
  primary: "#0BB589",
  primaryLight: "#E4F7ED",
  secondary: "#EF844C",
  background: "#F7FBFC",
  error: "#B12B2C",
  success: "#219653",
  info: "#00CFE8",
  overlay: "#696969",
  text: "#5C605E",
  secondaryText: "#82868B",
  disabled: "#B9BBC3",
  progress: "#E0E4E5",
  inputText: "#EFEFEF",
  card: "#F7F9FD",
  white: "#FFFFFF",
  black: "#000000",
};

export const SIZES = {
  borderRadius: 16,
};
