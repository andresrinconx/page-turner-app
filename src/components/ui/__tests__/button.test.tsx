/* eslint-disable prettier/prettier */
import { render, fireEvent } from "@testing-library/react-native";
import { View } from "react-native";
import Button from "../button";
import { COLORS } from "../../../lib/constants";

describe("Button", () => {
  it("renders with default props", () => {
    const { getByText } = render(<Button>Click me</Button>);
    const button = getByText("Click me");
    expect(button).toBeTruthy();
  });

  it("renders different variants correctly", () => {
    const variants = ["primary", "outline", "ghost"] as const;

    variants.forEach(variant => {
      const { getByTestId } = render(
        <Button variant={variant} testID="test-button">
          Test Button
        </Button>,
      );
      const button = getByTestId("test-button");
      expect(button.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining(
            variant === "primary"
              ? { backgroundColor: COLORS.primary, borderWidth: 0 }
              : variant === "outline"
                ? {
                  backgroundColor: "transparent",
                  borderWidth: 1,
                  borderColor: COLORS.primary,
                }
                : { backgroundColor: "transparent", borderWidth: 0 },
          ),
        ]),
      );
    });
  });

  it("renders different sizes correctly", () => {
    const sizes = ["sm", "md", "lg", "icon"] as const;

    sizes.forEach(size => {
      const { getByTestId } = render(
        <Button size={size} testID="test-button">
          Size Test
        </Button>,
      );
      const button = getByTestId("test-button");
      expect(button.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining(
            size === "sm"
              ? { paddingHorizontal: 12, paddingVertical: 6 }
              : size === "md"
                ? { paddingHorizontal: 16, paddingVertical: 8 }
                : size === "lg"
                  ? { paddingHorizontal: 24, paddingVertical: 12 }
                  : { padding: 8 },
          ),
        ]),
      );
    });
  });

  it("handles loading state correctly", () => {
    const { getByTestId, queryByText } = render(
      <Button loading testID="test-button">
        Loading Button
      </Button>,
    );

    expect(queryByText("Loading Button")).toBeNull();
    expect(getByTestId("test-button")).toBeTruthy();
  });

  it("handles disabled state correctly", () => {
    const onPress = jest.fn();
    const { getByTestId } = render(
      <Button disabled onPress={onPress} testID="test-button">
        Disabled Button
      </Button>,
    );

    const button = getByTestId("test-button");
    expect(button.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          opacity: 0.5,
        }),
      ]),
    );

    fireEvent.press(button);
    expect(onPress).not.toHaveBeenCalled();
  });

  it("handles fullWidth prop correctly", () => {
    const { getByTestId } = render(
      <Button fullWidth testID="test-button">
        Full Width
      </Button>,
    );
    const button = getByTestId("test-button");
    expect(button.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          width: "100%",
        }),
      ]),
    );
  });

  it("handles press events correctly", () => {
    const onPress = jest.fn();
    const { getByTestId } = render(
      <Button onPress={onPress} testID="test-button">
        Press Me
      </Button>,
    );
    const button = getByTestId("test-button");

    fireEvent.press(button);
    expect(onPress).toHaveBeenCalledTimes(1);
  });

  it("renders children correctly when not a string", () => {
    const { getByTestId } = render(
      <Button testID="test-button">
        <View testID="custom-child" />
      </Button>,
    );
    expect(getByTestId("custom-child")).toBeTruthy();
  });
});
