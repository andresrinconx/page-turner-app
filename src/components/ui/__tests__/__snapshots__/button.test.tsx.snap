// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot with custom props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 16,
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "center",
      },
      {
        "backgroundColor": "transparent",
        "borderColor": "#0BB589",
        "borderWidth": 1,
      },
      {
        "paddingHorizontal": 24,
        "paddingVertical": 12,
      },
      {
        "width": "100%",
      },
      {
        "opacity": 0.5,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "fontFamily": "System",
        },
        {
          "fontSize": 14,
          "lineHeight": 20,
        },
        {
          "textAlign": "left",
        },
        {
          "color": "#0BB589",
          "fontWeight": "500",
        },
        undefined,
      ]
    }
  >
    Custom Snapshot
  </Text>
</View>
`;

exports[`Button matches snapshot with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 16,
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#0BB589",
        "borderWidth": 0,
      },
      {
        "paddingHorizontal": 16,
        "paddingVertical": 8,
      },
      {
        "width": "auto",
      },
      {
        "opacity": 1,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "fontFamily": "System",
        },
        {
          "fontSize": 14,
          "lineHeight": 20,
        },
        {
          "textAlign": "left",
        },
        {
          "color": "#FFFFFF",
          "fontWeight": "500",
        },
        undefined,
      ]
    }
  >
    Snapshot Test
  </Text>
</View>
`;
